core/go/internal/sequencer/originator/transaction/state_machine.go

```diff
@@ -0,0 +1,586 @@
+	Action Action
+	If     Guard
+}
+
+type EventHandler struct {
+	Validator   func(ctx context.Context, txn *Transaction, event common.Event) (bool, error) // function to validate whether the event is valid for the current state of the transaction.  This is optional.  If not defined, the event is always considered valid.
+	Actions     []ActionRule                                                                  // list of actions to be taken when this event is received.  These actions are run before any transition specific actions
+	Transitions []Transition                                                                  // list of transitions that this event could trigger.  The list is ordered so the first matching transition is the one that will be taken.
+}
+
+type StateDefinition struct {
+	OnTransitionTo Action                     // function to be invoked when transitioning into this state.  This is invoked after any transition specific actions have been invoked
+	Events         map[EventType]EventHandler // rules to define what events apply to this state and what transitions they trigger.  Any events not in this list are ignored while in this state.
+}
+
+var stateDefinitionsMap map[State]StateDefinition
+
+func init() {
+	// Initialize state definitions in init function to avoid circular dependencies
+	stateDefinitionsMap = map[State]StateDefinition{
+		State_Initial: {
+			Events: map[EventType]EventHandler{
+				Event_Created: {
+					Transitions: []Transition{
+						{
+							To: State_Pending,
+						},
+					},
+				},
+			},
+		},
+		State_Pending: {
+			Events: map[EventType]EventHandler{
+				Event_Delegated: {
+					Transitions: []Transition{
+						{
+							To: State_Delegated,
+						},
+					},
+				},
+			},
+		},
+		State_Delegated: {
+			Events: map[EventType]EventHandler{
+				Event_AssembleRequestReceived: {
+					Validator: validator_AssembleRequestMatches,
+					Transitions: []Transition{
+						{
+							To: State_Assembling,
+						},
+					},
+				},
+				Event_CoordinatorChanged: {},
+			},
+		},
+		State_Assembling: {
+			OnTransitionTo: action_AssembleAndSign,
+			Events: map[EventType]EventHandler{
+				Event_AssembleAndSignSuccess: {
+					Transitions: []Transition{
+						{
+							To: State_Endorsement_Gathering,
+							On: action_SendAssembleSuccessResponse,
+						},
+					},
+				},
+				Event_AssembleRevert: {
+					Transitions: []Transition{
+						{
+							To: State_Reverted,
+							On: action_SendAssembleRevertResponse,
+						},
+					},
+				},
+				Event_AssemblePark: {
+					Transitions: []Transition{
+						{
+							To: State_Parked,
+							On: action_SendAssembleParkResponse,
+						},
+					},
+				},
+				Event_CoordinatorChanged: {
+					//would be very strange to have missed a bunch of heartbeats and switched coordinators if we recently received an assemble request but it is possible so we need to handle it
+					Transitions: []Transition{
+						{
+							To: State_Delegated,
```

Comment:
Feels very odd to go straight to `State_Delegated` here - surely we go to `State_Pending` (although that's under discussion) so we can send the delegation off again.

Or maybe this transition has an `On` function in reality that performs `reDelegate`